nmosauth/auth_server/token_generator.py

```diff
@@ -13,62 +13,62 @@
 # limitations under the License.
 
 import os
-from authlib.jose import jwt
+import re
 import datetime
+from authlib.jose import jwt
+
 from .oauth2 import authorization
 from .constants import NMOSAUTH_DIR, PRIVKEY_FILE
 
+ALLOWED_SCOPES = ["registration", "node", "query", "connection"]
+
 
 class TokenGenerator():
 
     def __init__(self):
         pass
 
-    def get_access_rights(self, user, scope):
-        if user is None:
-            return "client_credentials"
-        else:
-            if scope == "is-04":
-                access = user.is04
-            elif scope == "is-05":
-                access = user.is05
-            else:
-                access = None
-            return access
-
-    def get_audience(self, scope, access):
-        audience = []
-        if access is not None:
-            if scope == "is-04":
-                audience = [
-                    "registry",
-                    "query"
-                ]
-            elif scope == "is-05":
-                audience = [
-                    "senders",
-                    "receivers"
-                ]
-        return audience
+    def get_audience(self, client):
+        # Using Regex
+        redirect_uri = client.get_default_redirect_uri()
+        pattern = re.compile(r'(?:https?://)?(?:www\.)?[a-zA-Z0-9-]+((?:\.[a-zA-Z]+)+)(/?.*)')
```

Comment:
Can urllib.parse not help here?

Reply:
I did a  quick test and `urlparse` doesn't parse correctly when a protocol isn't specified (such as when requesting the FQDN using the socket library, such as in the auth server).. using this approach seems to get more consistent results with hostnames. I'd also likely end up doing some kind of parsing on the returned "netloc" result anyway so this seemed best

Reply:
I did a quick test and urlparse doesn't parse correctly when a protocol isn't specified (such as when requesting the FQDN using the socket library, such as in the auth server).. using this approach seems to get more consistent results with hostnames. I'd also likely end up doing some kind of parsing on the returned "netloc" result anyway so this seemed best

Reply:
Ok. I'd still perhaps be tempted to add the protocol where it didn't exist and use the built in library. I'll leave it up to you though

Reply:
Probably a better point is... what should we be basing the wildcarded `aud` URI claim on? I've used the `redirect_uri` here as we know the redirect URI is where we will be sending the auth code to, so the client would have no own that endpoint to use it. Originally I used the `client_uri` but there's nothing to stop a rogue client from registering with any old URI that the auth server will determine as valid, so I settled on not using that. The other option is to use the auth server's domain, which is the most trusted, but I wasn't sure if this would work if the auth server were operating in a different / higher zone (e.g. `broadcastcentre.org` as opposed to `camera.studio1.broadcastcentre.org`.)

Reply:
It may be that this needs to be configurable. In a more complex setup you might have Nodes which are only permitted access to `*.studio1.broadcastcentre.org`, but the broadcast controller needs access to `*.broadcastcentre.org`. Perhaps the auth server has a default, but it could be re-configured by an administrator against each registered client?

In the longer term https://tools.ietf.org/wg/oauth/draft-ietf-oauth-resource-indicators/ may help, and if a Node can see the domain the visible registries are in it could request an appropriate 'aud' claim in its token request. I'm uncertain whether we should go as far as to include this sort of thing in the spec though given it's not (quite) an RFC yet.